core/domain/suggestion_registry.py

```diff
@@ -0,0 +1,332 @@
+            self, suggestion_id, target_id, target_version_at_submission,
+            status, author_id, assigned_reviewer_id, final_reviewer_id,
+            change_cmd, score_category):
+        """Initializes a Suggestion object of type
+        SUGGESTION_TYPE_EDIT_STATE_CONTENT.
+        """
+        self.suggestion_id = suggestion_id
+        self.suggestion_type = (
+            suggestion_models.SUGGESTION_TYPE_EDIT_STATE_CONTENT)
+        self.target_type = suggestion_models.TARGET_TYPE_EXPLORATION
+        self.target_id = target_id
+        self.target_version_at_submission = target_version_at_submission
+        self.status = status
+        self.author_id = author_id
+        self.assigned_reviewer_id = assigned_reviewer_id
+        self.final_reviewer_id = final_reviewer_id
+        self.change_cmd = change_cmd
+        self.score_category = score_category
+
+    def validate(self):
+        """Validates a suggestion object of type
+        SUGGESTION_TYPE_EDIT_STATE_CONTENT.
+
+        Raises:
+            ValidationError: One or more attributes of the
+                SuggestionEditStateContent object are invalid.
+        """
+        super(SuggestionEditStateContent, self).validate()
+
+        if self.get_score_type() != suggestion_models.SCORE_TYPE_CONTENT:
+            raise utils.ValidationError(
+                'Expected the first part of score_category to be %s '
+                ', received %s' % (
+                    suggestion_models.SCORE_TYPE_CONTENT,
+                    self.get_score_type()))
+
+        if self.get_score_sub_type() not in constants.ALL_CATEGORIES:
+            raise utils.ValidationError(
+                'Expected the second part of score_category to be a valid'
+                ' category, received %s' % self.get_score_sub_type())
+
+        if 'cmd' not in self.change_cmd:
+            raise utils.ValidationError(
+                'Expected change_cmd to contain a cmd key')
+
+        if self.change_cmd['cmd'] != exp_domain.CMD_EDIT_STATE_PROPERTY:
+            raise utils.ValidationError(
+                'Expected cmd to be %s, received %s' % (
+                    exp_domain.CMD_EDIT_STATE_PROPERTY, self.change_cmd['cmd']))
+
+        if 'property_name' not in self.change_cmd:
+            raise utils.ValidationError(
+                'Expected change_cmd to contain a property_name key')
+
+        if (
+                self.change_cmd['property_name'] !=
+                exp_domain.STATE_PROPERTY_CONTENT):
+            raise utils.ValidationError(
+                'Expected property_name to be %s, received %s' % (
+                    exp_domain.STATE_PROPERTY_CONTENT,
+                    self.change_cmd['property_name']))
+
+        if 'state_name' not in self.change_cmd:
+            raise utils.ValidationError(
+                'Expected change_cmd to contain a state_name key')
+
+        states = exp_services.get_exploration_by_id(self.target_id).states
```

Comment:
Drop this check, otherwise if the state name changes and you make some subsequent change to the suggestion (e.g. change the reviewer ID) you won't be able to save it. Instead, do this check at the point of suggestion acceptance only (maybe have a separate method for that). Also make sure (going forward) to figure out how to deal with obsolete state names somehow in the suggestion UI.

Reply:
If the state name changes, and the state the suggestion was made to is invalid, shouldn't we invalidate it? That was my idea behind this check. Later we can try to convert invalid suggestions to valid ones by letting the author choose which state he meant to suggest to(post rename of state). But in case the state was deleted itself, the suggestion is truly invalid.

So if a suggestion is invalid, shouldn't we set status to invalid ASAP to prevent reviews and changes and then realising that the suggestion can't be accepted?

Reply:
But the trigger for the state name becoming invalid is the exploration commit, not a change to the suggestion. The effect of what you're doing here is just to prevent the suggestion from being edited further. (Right?)

Reply:
That is true. If validate() fails with an exception, then the suggestion cannot be accepted. I think it is logical to also prevent further edits to a suggestion if it is invalid. However, another possibility is in future, if we design a mechanism to make invalid suggestions valid (I feel this is similar to resolving merge conflicts!) then we don't want to prevent further edits and rather just validate just before accepting and if invalid, allow the author to edit in order to fix these issues.

Reply:
OK, but for now, maybe just have a message in frontend saying "This suggestion cannot be accepted because ..." and gray out the accept button, leaving only the reject button?

Reply:
Moved this to pre_accept_validate()